### PR TITLE
Tweaks to configure generation and testing

### PR DESCRIPTION
**Pull request by dra27**
Apropos of the discussion started in https://github.com/ocaml/ocaml/pull/2318#issuecomment-473852615

This GPR does three things (some, all, or even none of which may be wanted):

 - It declares `configure` to be a binary file. This means it doesn't appear in diffs (other than to say it's changed) which removes the temptation to comment on its output in code review (note that you can view the actual diff by issuing `git diff --text configure`)
 - It hooks the Travis logic used for `check-typo` to verify that `configure.ac` still generates `configure`. `configure` should in many ways be treated as the `boot/` artefacts.
 - It switches Travis from Ubuntu 14.04 LTS to Ubuntu 16.04 LTS

The most common "oscillation" as noted in the original discussion is the runstate patch. All distributions should be using autoconf 2.69, as it was released 7 years ago. However, Debian 8 (Jessie) and Ubuntu 14.04 LTS both lack the runstate patch (as does Cygwin's autoconf). I think RHEL may also lack it.

It would be possible to adapt the test not to care about the runstate difference - although, having used Cygwin's autoconf for quite a few years with this, I tend just to use `git checkout -p` to get rid of the `runstate` deletion. We don't use that flag, so it doesn't particularly matter.

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
BTW, what do you think about adding a check in `autogen` after running for autoconf and display a warning if we can't find the runstate stuff?

**Comment by dra27** [comment]
@diml - I was thinking about this yesterday, given the lack of this patch on RedHat-based Linuxes (which I believe is what you probably saw?). I don't have access to a RHEL-box to verify it, but the problem with warning is what does the developer do then?

I've pushed an extra commit which eliminates the `runstatedir` change from `configure` for consideration.

Just for some background - this patch has been in autoconf since 2013 (see [their git repo](http://git.savannah.gnu.org/cgit/autoconf.git/commit/?id=a197431414088a417b407b9b20583b2e8f7363bd) - it's even a RedHat patch!) but despite all the activity on their master branch, they've not done a release.

**Comment by dra27** [comment]
Oh, the test for autoconf 2.69 I've added since autoconf embeds its version number in the `configure` file - but it's inconceivable that anyone should want to be using an older autoconf.

**Comment by ghost** [comment]
I was thinking that a warning is better than the current silent behavior. But maybe we are blowing this out of proportion. At the end of the day, we are not going to modify the configure.ac file all the time. I don't even know what this runstate block is doing, but my guess is that it is not very important?

**Comment by dra27** [comment]
The `--runstatedir` is about changing the default location of `/var/run` to `/run` (or maybe vice versa?!)… so I really can't see that it's something we'd ever need! I agree that a warning about the removal of it from the script is probably overkill.

The only thing which is important is to be sure that `configure` in a given PR was generated from `configure.ac` (the most common thing is to forget to rebuild it, cf. https://github.com/ocaml/ocaml/pull/8528#discussion_r267255713)

Can I go ahead and merge?

File: tools/ci/travis/travis-ci.sh [hunk 1]
```diff
@@ -223,8 +223,23 @@ CheckTypoTree () {
...
   rm -f tmp-index
+  if [ -e CHECK_CONFIGURE ] ; then
+    rm -f CHECK_CONFIGURE
+    echo "configure or configure.ac altered in $1"
+    echo "Verifying that configure.ac generates configure"
+    git checkout "$1"
+    mv configure configure.ref
+    ./autogen
+    if ! diff -q configure configure.ref >/dev/null ; then
+      echo "configure.ac no longer generates configure"
```
Comment (ghost) [inline tools/ci/travis/travis-ci.sh]:
```suggestion
      echo "configure.ac no longer generates configure, please run ./autogen and commit"
```

**Comment by ghost** [comment]
Seems good to me. I just made a suggestion to improve the error message

File: tools/ci/travis/travis-ci.sh (hunk 1 above)
Comment (dra27) [reply tools/ci/travis/travis-ci.sh]: Hah - we need GitHub to run `check-typo` on suggestions :wink:

This is a good idea, I just had to reformat it for line-length.

**Comment by dra27** [comment]
I'll just let the `check-typo` Travis job run one more time to be sure and then merge

**Review by ghost — APPROVED** [review]

**Comment by shindere** [comment]
Sorry for being late on that one, but, @dra27, in commit
1a7da65d8d026a27df5a7ebda7d0875ef708760f, is there a particular reason
for not having used the `AC_PREREQ` autoconf macro, which seems to
do exactly what your script does, but in a more idiomatic way?

**Comment by dra27** [comment]
@shindere - because I didn't know about it :o)

**Comment by shindere** [comment]
David Allsopp (2019/06/10 00:44 -0700):
> @shindere - because I didn't know about it :o)

Okay no problem! Will you do a PR to simplify this?

If you do so please do not hesitate to assign it to me as a reviewer.

**Comment by dra27** [comment]
Just about to submit it!

**Pull request merged**